src/main/java/RecommendationsServlet.java

```diff
@@ -0,0 +1,184 @@
+    Arrays.asList("under-18", "over-18", "all-ages", "male", "female"));
+  private ArrayList<String> action = new ArrayList<String>(
+    Arrays.asList("live", "work", "moved"));
+  private ArrayList<String> location = new ArrayList<String>(
+    Arrays.asList("state", "01", "02", "04", "05", "06", "08",
+    "09", "10", "11", "12", "13", "15", "16", "17", "18", "19",
+    "20", "21", "22", "23", "24", "25", "26", "27", "28", "29",
+    "30", "31", "32", "33", "34", "35", "36", "37", "38", "39",
+    "40", "41", "42", "44", "45", "46", "47", "48", "49", "50",
+    "51", "53", "54", "55", "56"));
+  private ArrayList<String> years = new ArrayList<String> (
+    Arrays.asList("2010", "2011", "2012", "2013", "2014", "2015", "2016", "2017", "2018")); 
+  private HashMap<String, Integer> personTypeMap = new HashMap<String, Integer>();
+  private HashMap<String, Integer> actionMap = new HashMap<String, Integer>();
+  private HashMap<String, Integer> locationMap = new HashMap<String, Integer>(); 
+  private HashMap<String, Integer> yearsMap = new HashMap<String, Integer>(); 
+
+  // if the user doesn't have any history to draw from to get recommendations, 
+  // return default recommendations that would be applicable to most users which is
+  // the number of children, adults, people, men, and women who lived in each 
+  // US state in 2018. 
+  private ArrayList<VisualizationData> getDefaultRecommendations(String userId) {
+    ArrayList<VisualizationData> defaultRecommendations = new ArrayList<VisualizationData>();
+    for (String person: personType) {
+      defaultRecommendations.add(new VisualizationData(userId, person, "live", "state", "2018"));
+    }
+    return defaultRecommendations;
+  }
+
+  // uses java.util.random to get a random integer between 
+  // 0 and upperBound-1 in order to access a random element 
+  // for each of the options
+  private int getRandomInt(int upperBound) {
+    Random randomObject = new Random();
+    int randomInt = randomObject.nextInt(upperBound);
+    return randomInt;
+  }
+
+  // CHeck to see if the current recommendation is valid by making sure that it isn't an impossible combination 
+  // and isn't already one that the user has viewed.
+  private boolean isRecommendationValid(VisualizationData recommendation, ArrayList<VisualizationData> userHistory) {
+      return (!(userHistory.contains(recommendation) || 
+        (recommendation.getPersonType() == "under-18" && recommendation.getAction() == "work") ||
+        (recommendation.getPersonType() == "under-18" && recommendation.getAction() == "moved"))); 
+  }
+
+  // Returns a VisualizationData whose contents are completely random
+  // in order to alleviate recommendation fatigue. 
+  private VisualizationData getRandomRecommendation(String userId, ArrayList<VisualizationData> userHistory) {
+    ArrayList<VisualizationData> invalidRecommendations = new ArrayList<VisualizationData>();
+    boolean isValid = false;
+    while (!isValid) {
```

Comment:
You can be stuck in this loop based on how random the integer generation actually is. This might be ok for now, but a cleaner solution might be to generate only valid combinations.

Reply:
Updated so the function iterates as long as the unuseableRecommendation array is not full of all possible recommendations. If a valid recommendation is generated, then it is returned immediately.

I don't think this fixes the problem bc invalid recommendations can still be generated